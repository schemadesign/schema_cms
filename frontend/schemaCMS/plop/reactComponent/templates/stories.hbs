import React from 'react';
import { storiesOf } from '@storybook/react';

import { {{ pascalCase name }} } from './{{ camelCase name }}.component';

export const defaultProps = {};

storiesOf('{{ pascalCase name }}', module).add('Default', () => <{{ pascalCase name }} {...defaultProps} />);
