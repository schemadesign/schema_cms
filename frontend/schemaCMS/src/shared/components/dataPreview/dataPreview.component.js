import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import { Button, Icons } from 'schemaUI';
import { isEmpty } from 'ramda';

import { Loader } from '../loader';
import { Table } from '../table';
import { FieldDetail } from '../fieldDetail';
import messages from './dataPreview.messages';
import {
  Container,
  Navigation,
  NavigationLabel,
  NavigationButton,
  Content,
  buttonStyles,
  arrowStyles,
} from './dataPreview.styles';
import { StepNavigation } from '../stepNavigation';
import { getTableData } from '../../utils/helpers';

const INITIAL_STEP = 0;

export default class DataPreview extends PureComponent {
  static propTypes = {
    fields: PropTypes.object,
    previewTable: PropTypes.array,
    fetchFields: PropTypes.func.isRequired,
    unmountFields: PropTypes.func.isRequired,
    intl: PropTypes.object.isRequired,
    dataSource: PropTypes.object.isRequired,
  };

  static getDerivedStateFromProps({ fields }, { isLoading, step }) {
    if (isLoading && !isEmpty(fields)) {
      const fieldsIds = Object.keys(fields);

      return {
        step,
        isLoading: false,
        countFields: fieldsIds.length,
        fieldsIds: [null, ...fieldsIds],
      };
    }

    return null;
  }

  state = {
    isLoading: true,
    step: INITIAL_STEP,
  };

  componentDidMount() {
    const { id: dataSourceId } = this.props.dataSource;

    this.props.fetchFields({ dataSourceId });
  }

  componentWillUnmount() {
    this.props.unmountFields();
  }

  getFieldDetailsData() {
    const { step, fieldsIds } = this.state;
    const id = fieldsIds[step];
    const data = this.props.fields[id];

    return { id, data, intl: this.props.intl };
  }

  handleNavigation = direction => {
    const { step, countFields } = this.state;
    const updatedStep = step + direction;

    if (updatedStep >= INITIAL_STEP && updatedStep <= countFields) {
      this.setState({
        step: updatedStep,
      });
    }
  };

  handlePreviewStep = () => this.handleNavigation(-1);

  handleNextStep = () => this.handleNavigation(1);

  renderNavigation() {
    const { step, countFields } = this.state;
    const { intl } = this.props;

    const label = step
      ? intl.formatMessage(messages.ofFields, { step, countFields })
      : intl.formatMessage(messages.fields, { countFields });

    const isPreviousDisabled = step === INITIAL_STEP;
    const isNextDisabled = step === countFields;

    return (
      <Navigation>
        <NavigationButton>
          <Button onClick={this.handlePreviewStep} disabled={isPreviousDisabled} customStyles={buttonStyles}>
            <Icons.ArrowLeftIcon customStyles={arrowStyles(isPreviousDisabled)} />
          </Button>
        </NavigationButton>
        <NavigationLabel>{label}</NavigationLabel>
        <NavigationButton>
          <Button onClick={this.handleNextStep} disabled={isNextDisabled} customStyles={buttonStyles}>
            <Icons.ArrowRightIcon customStyles={arrowStyles(isNextDisabled)} />
          </Button>
        </NavigationButton>
      </Navigation>
    );
  }

  renderTable = props => <Table {...props} numberedRows />;

  renderFieldDetails = props => <FieldDetail {...props} />;

  renderContent() {
    const { step } = this.state;
    const tableData = getTableData(this.props.previewTable);
    const content = step ? this.renderFieldDetails(this.getFieldDetailsData()) : this.renderTable(tableData);

    return (
      <Content>
        {this.renderNavigation()}
        {content}
      </Content>
    );
  }

  render() {
    const content = this.state.isLoading ? <Loader /> : this.renderContent();

    return (
      <Container>
        <Helmet title={this.props.intl.formatMessage(messages.pageTitle)} />
        {content}
        <StepNavigation {...this.props} />
      </Container>
    );
  }
}
