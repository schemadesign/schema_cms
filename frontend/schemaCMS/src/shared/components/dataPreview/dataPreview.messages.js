/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  pageTitle: {
    id: 'shared.components.dataPreview.pageTitle',
    defaultMessage: 'Fields',
  },
  fields: {
    id: 'shared.components.dataPreview.fields',
    defaultMessage: '{countFields} {countFields, plural,one {Field} other {Fields}}',
  },
  ofFields: {
    id: 'shared.components.dataPreview.ofFields',
    defaultMessage: '{step} of {countFields} {countFields, plural, one {Field} other {Fields}}',
  },
});
