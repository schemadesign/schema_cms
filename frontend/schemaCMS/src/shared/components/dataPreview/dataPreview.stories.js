import React from 'react';
import { storiesOf } from '@storybook/react';
import { Theme } from 'schemaUI';

import { withTheme } from '../../../.storybook/decorators';
import DataPreview from './dataPreview.component';
import { STATUS_DONE } from '../../../modules/dataSource/dataSource.constants';

const previewTable = [
  {
    'First Name': 'George',
    'Last Name': 'Washington',
    'Full Name': 'George Washington',
    Birthplace: 'Westmoreland County',
    'State of Birth': 'VA',
    Birthday: '02/22/1732',
    Url: 'some.url.com',
  },
  {
    'First Name': 'John',
    'Last Name': 'Adams',
    'Full Name': 'John Adams',
    Birthplace: 'Braintree',
    'State of Birth': 'MA',
    Birthday: '10/30/1735',
    Url: 'some.url.com',
  },
  {
    'First Name': 'Thomas',
    'Last Name': 'Jefferson',
    'Full Name': 'Thomas Jefferson',
    Birthplace: 'Shadwell',
    'State of Birth': 'VA',
    Birthday: '04/13/1743',
    Url: 'some.url.com',
  },
  {
    'First Name': 'James',
    'Last Name': 'Madison',
    'Full Name': 'James Madison',
    Birthplace: 'Port Conway',
    'State of Birth': 'VA',
    Birthday: '03/16/1751',
    Url: 'some.url.com',
  },
  {
    'First Name': 'James',
    'Last Name': 'Monroe',
    'Full Name': 'James Monroe',
    Birthplace: 'Monroe Hall',
    'State of Birth': 'VA',
    Birthday: '04/28/1758',
    Url: 'some.url.com',
  },
];

const fields = {
  'First Name': {},
  'Last Name': {},
  'Full Name': {},
  Birthplace: {},
  'State of Birth': {},
  Birthday: {},
  Url: {},
};

const dataSource = {
  id: 1,
  project: 1,
  status: STATUS_DONE,
};

export const defaultProps = {
  fields,
  previewTable,
  fetchFields: Function.prototype,
  unmountFields: Function.prototype,
  match: {
    params: {
      step: '2',
    },
  },
  history: {
    push: Function.prototype,
  },
  intl: {
    formatMessage: ({ defaultMessage }) => defaultMessage,
  },
  dataSource,
};

storiesOf('DataSource/View/Fields', module)
  .addDecorator(withTheme(Theme.light))
  .add('Default', () => <DataPreview {...defaultProps} />);
