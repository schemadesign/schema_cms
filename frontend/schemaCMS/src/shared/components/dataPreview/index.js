export { default as DataPreview } from './dataPreview.component';
