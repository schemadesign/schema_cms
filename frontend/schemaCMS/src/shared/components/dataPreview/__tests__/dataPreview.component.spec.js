import React from 'react';
import { shallow } from 'enzyme';

import DataPreview from '../dataPreview.component';
import { defaultProps } from '../dataPreview.stories';

describe('DataPreview: Component', () => {
  const component = props => <DataPreview {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });
});
