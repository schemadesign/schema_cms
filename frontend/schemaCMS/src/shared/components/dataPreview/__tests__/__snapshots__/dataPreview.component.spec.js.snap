// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataPreview: Component should render correctly 1`] = `
<styled.div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="Fields"
  />
  <styled.div>
    <styled.div>
      <styled.div>
        <Component
          customStyles={
            Object {
              "height": 60,
              "width": 60,
            }
          }
          disabled={true}
          onClick={[Function]}
        >
          <Component
            customStyles={
              Object {
                "opacity": 0.3,
              }
            }
          />
        </Component>
      </styled.div>
      <styled.div>
        {countFields} {countFields, plural,one {Field} other {Fields}}
      </styled.div>
      <styled.div>
        <Component
          customStyles={
            Object {
              "height": 60,
              "width": 60,
            }
          }
          disabled={false}
          onClick={[Function]}
        >
          <Component
            customStyles={Object {}}
          />
        </Component>
      </styled.div>
    </styled.div>
    <Table
      header={
        Array [
          "First Name",
          "Last Name",
          "Full Name",
          "Birthplace",
          "State of Birth",
          "Birthday",
          "Url",
        ]
      }
      numberedRows={true}
      rows={
        Array [
          Array [
            "George",
            "Washington",
            "George Washington",
            "Westmoreland County",
            "VA",
            "02/22/1732",
            "some.url.com",
          ],
          Array [
            "John",
            "Adams",
            "John Adams",
            "Braintree",
            "MA",
            "10/30/1735",
            "some.url.com",
          ],
          Array [
            "Thomas",
            "Jefferson",
            "Thomas Jefferson",
            "Shadwell",
            "VA",
            "04/13/1743",
            "some.url.com",
          ],
          Array [
            "James",
            "Madison",
            "James Madison",
            "Port Conway",
            "VA",
            "03/16/1751",
            "some.url.com",
          ],
          Array [
            "James",
            "Monroe",
            "James Monroe",
            "Monroe Hall",
            "VA",
            "04/28/1758",
            "some.url.com",
          ],
        ]
      }
    />
  </styled.div>
  <StepNavigation
    fetchFields={[Function]}
    fields={
      Object {
        "Birthday": Object {},
        "Birthplace": Object {},
        "First Name": Object {},
        "Full Name": Object {},
        "Last Name": Object {},
        "State of Birth": Object {},
        "Url": Object {},
      }
    }
    intl={
      Object {
        "formatMessage": [Function],
      }
    }
    match={
      Object {
        "params": Object {
          "dataSourceId": "1",
          "projectId": "1",
        },
      }
    }
    previewTable={
      Array [
        Object {
          "Birthday": "02/22/1732",
          "Birthplace": "Westmoreland County",
          "First Name": "George",
          "Full Name": "George Washington",
          "Last Name": "Washington",
          "State of Birth": "VA",
          "Url": "some.url.com",
        },
        Object {
          "Birthday": "10/30/1735",
          "Birthplace": "Braintree",
          "First Name": "John",
          "Full Name": "John Adams",
          "Last Name": "Adams",
          "State of Birth": "MA",
          "Url": "some.url.com",
        },
        Object {
          "Birthday": "04/13/1743",
          "Birthplace": "Shadwell",
          "First Name": "Thomas",
          "Full Name": "Thomas Jefferson",
          "Last Name": "Jefferson",
          "State of Birth": "VA",
          "Url": "some.url.com",
        },
        Object {
          "Birthday": "03/16/1751",
          "Birthplace": "Port Conway",
          "First Name": "James",
          "Full Name": "James Madison",
          "Last Name": "Madison",
          "State of Birth": "VA",
          "Url": "some.url.com",
        },
        Object {
          "Birthday": "04/28/1758",
          "Birthplace": "Monroe Hall",
          "First Name": "James",
          "Full Name": "James Monroe",
          "Last Name": "Monroe",
          "State of Birth": "VA",
          "Url": "some.url.com",
        },
      ]
    }
    unmountFields={[Function]}
  />
</styled.div>
`;
